src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs

```diff
@@ -46,28 +47,52 @@ public class PageActionDescriptorChangeProvider : IActionDescriptorChangeProvide
             rootDirectory = rootDirectory.TrimEnd('/');
 
             var importFileAtPagesRoot = rootDirectory + "/" + templateEngine.Options.ImportsFileName;
-            _additionalFilesToTrack = templateEngine.GetImportItems(importFileAtPagesRoot)
-                .Select(item => item.FilePath)
-                .ToArray();
+            var additionalImportFilePaths = templateEngine.GetImportItems(importFileAtPagesRoot)
+                .Select(item => item.FilePath);
+            var pagesRootSearchPattern = rootDirectory + "/**/*.cshtml";
 
-            _searchPattern = rootDirectory + "/**/*.cshtml";
+            if (razorPagesOptions.Value.EnableAreas)
+            {
+                var areaRootDirectory = razorPagesOptions.Value.AreaRootDirectory;
+                Debug.Assert(!string.IsNullOrEmpty(areaRootDirectory));
+                areaRootDirectory = areaRootDirectory.TrimEnd('/');
+
+                var importFileAtAreaPagesRoot = areaRootDirectory + "/" + templateEngine.Options.ImportsFileName;
+                var importPathsOutsideAreaPagesRoot = templateEngine.GetImportItems(importFileAtAreaPagesRoot)
+                    .Select(item => item.FilePath);
+
+                additionalImportFilePaths = additionalImportFilePaths
+                    .Concat(importPathsOutsideAreaPagesRoot)
+                    .Distinct(StringComparer.OrdinalIgnoreCase);
+
+                _searchPatterns = new[]
+                {
+                    pagesRootSearchPattern,
+                    areaRootDirectory + "/**/*.cshtml",
+                };
+            }
+            else
+            {
+                _searchPatterns = new[] { pagesRootSearchPattern, };
+            }
+
+            _additionalFilesToTrack = additionalImportFilePaths.ToArray();
```

Comment:
Would it be possible to expose this and cover it in a unit test? I think it would be really helpful to see a test that shows exactly what gets tracked.

Reply:
I checked the tests, and this seems like it has enough coverage. I thought this code was a bit hard to follow, even as someone who already knows it pretty well. Consider cleaning this up or adding some explanations in comments